qdr/_ranker.cc

```diff
@@ -0,0 +1,416 @@
+
+#include <math.h>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <stdexcept>
+
+
+typedef std::unordered_map<std::string, std::pair<uint64_t, uint64_t> >
+    counts_t;
+
+
+// count of word occurances in a doc
+typedef std::unordered_map<std::string, uint64_t> word_counts_t;
+
+// a tokenized document
+typedef std::vector<std::string> doc_t;
+
+word_counts_t count_words(const doc_t& document)
+{
+    /// Given a vector of tokens, return a map token -> count in vector
+    word_counts_t ret(document.size() * 3);
```

Comment:
ret an unordered_map so this is a quick and dirty attempt at setting the hash sparsity initially to avoid rehashes.  I took a guess that a final hash sparsity of 10% would be a good tradeoff between space/efficiency.  The number of keys in the final hash is the number of unique words which is about 1/3 the number of total words in the page is checked.  So making a hash size of 3*words in document gives about a 10% sparsity.